internal/velero/storagelocation.go

```diff
@@ -0,0 +1,131 @@
+	validationFrequency := p.DefaultStoreValidationFrequency
+	// If the bsl validation frequency is not specifically set, skip this block and continue, using the server's default
+	if location.Spec.ValidationFrequency != nil {
+		validationFrequency = location.Spec.ValidationFrequency.Duration
+		if validationFrequency == 0 {
+			log.Debug("Validation period for this backup location is set to 0, skipping validation")
+			return false
+		}
+
+		if validationFrequency < 0 {
+			log.Debugf("Validation period must be non-negative, changing from %d to %d", validationFrequency, p.DefaultStoreValidationFrequency)
+			validationFrequency = p.DefaultStoreValidationFrequency
+		}
+	}
+
+	lastValidation := location.Status.LastValidationTime
+	if lastValidation != nil { // always ready to validate the first time around, so only even do this check if this has happened before
+		nextValidation := lastValidation.Add(validationFrequency) // next validation time: last validation time + validation frequency
+		if time.Now().UTC().Before(nextValidation) {              // ready only when NOW is equal to or after the next validation time
+			return false
+		}
+	}
+
+	return true
+}
+
+// IsValidFor verifies if a storage is valid for a given backup storage location.
+func (p *StorageLocation) IsValidFor(location *velerov1api.BackupStorageLocation, log logrus.FieldLogger) error {
+	pluginManager := p.NewPluginManager(log)
+	defer pluginManager.CleanupClients()
+
+	backupStore, err := p.NewBackupStore(location, pluginManager, log)
+	if err != nil {
+		return err
+	}
+
+	if err := backupStore.IsValid(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// UpdatePhase patches the status.phase field as well as the status.lastValidationTime to the current time
+func (p *StorageLocation) UpdatePhase(location *velerov1api.BackupStorageLocation, phase velerov1api.BackupStorageLocationPhase) error {
+	statusPatch := client.MergeFrom(location.DeepCopyObject())
+	location.Status.Phase = phase
+	location.Status.LastValidationTime = &metav1.Time{Time: time.Now().UTC()}
+	if err := p.Client.Status().Patch(p.Ctx, location, statusPatch); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// BackupStorageLocationsExist verifies if there are any backup storage locations.
+// For all purposes, if either there is an error while attempting to fetch items or
+// if there are no items an error would be returned since the functioning of the system
+// would be haulted.
+func BackupStorageLocationsExist(kbClient client.Client, ctx context.Context, namespace string) (velerov1api.BackupStorageLocationList, error) {
+	var locations velerov1api.BackupStorageLocationList
+	if err := kbClient.List(ctx, &locations, &client.ListOptions{
+		Namespace: namespace,
+	}); err != nil {
+		return velerov1api.BackupStorageLocationList{}, err
+	}
+
+	if len(locations.Items) == 0 {
+		return velerov1api.BackupStorageLocationList{}, errors.New("no backup storage locations found")
```

Comment:
I don't recall discussing this before but in this context this is an error, ie, not having any BSL is a halting error.

Reply:
I know the original content of #1967 was requesting that Velero not halt when there were no BSLs. That's because the reporter's scenario was that they created BSLs on-demand; they had a higher-order system that received a backup request, created a BSL, then issued the backup to Velero.

As it stands now (without this PR merged), if you have a BSL and get the Velero server running, then delete the BSL, the Velero server does not halt, because the check for BSLs being present only happened on startup.

#2382 tried to unify the behavior into a liveness check that relied on Kubernetes to restart Velero's server until a BSL was created, though at this point I don't think that's a good solution to the problem.

I think having Velero stay up when there are no BSLs found, and retry until there are, makes it more resilient when it is used as part of a larger system. However, if we can error here and cause the reconciliation loop to keep retrying, then I think we accomplish the goal and keep the Velero server up.

Reply:
Re-reading the code, I think the usage makes sense.

It's an error from this function, but the error is viewed, logged, and then moved past in the controller.

Reply:
Yes, and we had discussed this behavior change as a group before, that's why I went ahead and implemented it. Also, there is a case where users pair up 1 backup with 1 bsl, and they want to create that bsl in a scripted way at backup time, and having the server not boot up because it failed validation was an issue.

The controller (in this PR) will continuously be reconciling, and will log an error whenever the system is in a state where there are no BSL or there are BSLs but they are all invalid.